src/geode/model/svg_input.cpp

```diff
@@ -0,0 +1,431 @@
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ */
+
+#include <geode/model/detail/svg_input.h>
+
+#include <fstream>
+#include <unordered_map>
+
+#include <pugixml.hpp>
+
+#include <geode/geometry/bounding_box.h>
+#include <geode/geometry/nn_search.h>
+#include <geode/geometry/point.h>
+#include <geode/geometry/vector.h>
+
+#include <geode/mesh/builder/edged_curve_builder.h>
+#include <geode/mesh/builder/point_set_builder.h>
+
+#include <geode/model/mixin/core/corner.h>
+#include <geode/model/mixin/core/line.h>
+#include <geode/model/representation/core/section.h>
+
+namespace
+{
+    static constexpr double FRACTION = 1e-5;
+
+    class SVGInputImpl
+    {
+    public:
+        SVGInputImpl( std::string filename, geode::Section& section )
+            : file_( std::move( filename ) ),
+              section_( section ),
+              builder_{ section }
+        {
+            OPENGEODE_EXCEPTION( file_.good(),
+                "[SVGInput] Error while opening file: " + filename );
+            auto ok = document_.load_file( filename.c_str() );
+            OPENGEODE_EXCEPTION(
+                ok, "[SVGInput] Error while parsing file: " + filename );
+        }
+
+        void read_file()
+        {
+            for( const auto& group : document_.child( "svg" ).children( "g" ) )
+            {
+                read_group_paths( group );
+            }
+        }
+
+        void process_paths()
+        {
+            for( auto& path : paths_ )
+            {
+                format_path( path );
+                auto tokens = split_path( path );
+                process_tokens( tokens );
+            }
+        }
+
+        void build_topology()
+        {
+            const auto epsilon = compute_epsilon();
+            const geode::NNSearch2D colocater( potential_corners_ );
+            const auto colocated_info =
+                colocater.colocated_index_mapping( epsilon );
+            const auto corner_ids = create_corners( colocated_info );
+            build_corner_line_relations( colocated_info, corner_ids );
+            create_line_unique_vertices();
+        }
+
+    private:
+        struct Command
+        {
+            Command() = default;
+
+            void update( const char token )
+            {
+                letter = std::tolower( token );
+                absolute = std::isupper( token );
+            }
+
+            void apply( geode::Point2D& cur_position,
```

Comment:
```suggestion
            geode::Point2D apply( const geode::Point2D& position,
```
Maybe easier to understand. Output parameter should be in the return as much as possible
position = command.apply(position, params);